passphrase.go

```diff
@@ -30,3 +88,193 @@ func GenerateKeyFromPassphraseAndCopy(passphrase string, key []byte) {
 	subtle.ConstantTimeCopy(1, key[:], tmpKey[:])
 	return
 }
+
+// getNonce parses the string representation of the key validation code and returns the contained nonce if found; if
+// there is no nonce in the code it sets the 'hasNonce' flag to false
+// 'code' format: hexadecimal_hmac[:base10_32bit_nonce]
+func getNonce(code string) (nonce uint32, hasNonce bool) {
+	hasNonce = false
+	if checksumLen := strings.IndexByte(code, Separator); checksumLen == -1 {
+		return
+	} else if tmp, err := strconv.Atoi(code[checksumLen+1:]); err != nil {
+		return
+	} else {
+		nonce = uint32(tmp)
+	}
+	hasNonce = true
+	return
+}
+
+// randomUint32 generates a random unsigned integer
+func randomUint32() (n uint32, err error) {
+	var buf [4]byte
+	if _, err = rand.Read(buf[:]); err != nil {
+		n = 0
+		return
+	}
+	n = 0
+	for i, b := range buf {
+		n = n | uint32(b)<<(8*uint32(i))
+	}
+	return
+}
+
+func registerHashFunctions() {
+	crypto.RegisterHash(crypto.SHA256, sha256.New)
+}
+
+// NewKeyValidator returns a key validator which be used either globally or in its own thread.
+// length indicates how much of the key checksum hexadecimal encoding is used for validation (0 < length <= 2*cryptoHash.Size())
+// flags: nonce | pre-allocated validator
+func NewKeyValidator(cryptoHash crypto.Hash, key []byte, length int, flags ...bool) (Validator, error) {
+	var (
+		nonce     uint32    = 0
+		nonceFlag bool      = false
+		mac       hash.Hash = nil
+		err       error     = nil
+	)
+	registerHashFunctions()
+	switch len(flags) {
+	case 0:
+		// nonce not used, on-the-fly validator
+		nonceFlag = false
+		mac = nil
+	case 1:
+		// nonce flag specified, on-the-fly validator
+		nonceFlag = flags[0]
+		mac = nil
+	case 2:
+		// nonce flag specified, pre-allocated validator flag specified
+		fallthrough
+	default:
+		nonceFlag = flags[0]
+		if flags[1] {
+			// pre-allocated validator
+			mac = hmac.New(cryptoHash.New, key)
+		} else {
+			// on-the-fly validator
+			mac = nil
+		}
+	}
+	if nonceFlag {
+		// generate a random nonce
+		if nonce, err = randomUint32(); err != nil {
+			return nil, err
+		}
+	}
+	if length < ChecksumMinLength {
+		length = ChecksumMinLength
+	} else if length > 2*cryptoHash.Size() {
+		length = 2 * cryptoHash.Size()
+	}
+	return &KeyValidator{key, cryptoHash, mac, length, nonceFlag, nonce}, nil
+}
+
+// computeWithNonce computes the validation code using an optional nonce specified as parameter
+func (vtor *KeyValidator) computeWithNonce(nonce ...uint32) (kv KeyValidation) {
+	mac := vtor.mac
+	if mac == nil {
+		// allocate an "ephemeral" hmac object
+		mac = hmac.New(vtor.hash.New, vtor.key)
+	}
+	mac.Reset()
+	mac.Write(vtor.key)
```

Comment:
Why not directly sha256.Sum256(vtor.key + none) ? I mean everything is very abstract now with interfaces and everything, but doesn't it introduce too much overhead?

Reply:
it is done using hmac https://en.wikipedia.org/wiki/HMAC / https://tools.ietf.org/html/rfc2104 which is the std. for computing message auth. codes. HMAC is abstracting the hash functions.

Reply:
there are 2 types of KeyValidators:
- initialize it once (which is basically about the initialization of the HMAC/hash functions) and reuse it; it has to be _thread local_ in a multithreading environment. so one KeyValidator per thread has to be initialized.
- the HMAC is initialized on the fly when it is used. this one is thread safe in the sense that a global one can be used in any thread at the cost of re-initializing the HMAC/hash function.